### PR TITLE
[Train][Doc] LightningTrainer end-2-end starter example [no_early_kickoff]

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

We need to provide end-to-end examples for users to onboard LightningTrainer. This is a starter example which uses
- simple linear model for MNIST dataset
- `pl.LightningDataModule` as input data source
- `pl.Trainer.test()` for evaluation

We'll also have an advanced example using Ray Data + BatchPredictor + Ray Tune later.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MNISTDataModule(pl.LightningDataModule):\n",
+    "    def __init__(self, batch_size=100):\n",
+    "        super().__init__()\n",
+    "        self.data_dir = os.getcwd()\n",
+    "        self.batch_size = batch_size\n",
+    "        self.transform = transforms.Compose(\n",
+    "            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]\n",
```
Comment (justinvyu) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:54]: Btw, is the batch size of a datamodule global batch size across all workers? Or batch size per worker?

What about for a torch DataLoader?

**Review by justinvyu — COMMENTED** [review]
Looks good!

Can we also add usage of the `LightningPredictor` at the end? As a separate path to `trainer.test`.

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+    "        )\n",
+    "\n",
+    "    def setup(self, stage=None):\n",
+    "        # split data into train and val sets\n",
+    "        mnist = MNIST(\n",
+    "            self.data_dir, train=True, download=True, transform=self.transform\n",
```
Comment (justinvyu) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:60]: Do we need a file lock for this downloading operation?

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+    "import torch.nn.functional as F\n",
+    "from torch.utils.data import DataLoader, random_split, Subset\n",
+    "from torchmetrics import Accuracy\n",
+    "from torchvision.datasets import MNIST\n",
+    "from torchvision import transforms\n",
+    "\n",
+    "import pytorch_lightning as pl\n",
+    "from pytorch_lightning import trainer\n",
+    "from pytorch_lightning.core import datamodule\n",
+    "from pytorch_lightning.loggers.csv_logs import CSVLogger"
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_mnist_example.ipynb:30]: Actually the default is a `tensorboardLogger`. Here we just want to demo how to properly use `LightningConfigBuilder`.

File: doc/source/conf.py [hunk 4]
```diff
@@ -341,7 +341,8 @@ def filter_out_undoc_class_members(member_name, class_name, module_name):
     "trainMlflow": "MLflow,Training",
     "trainTuneTensorflow": "TensorFlow,Training,Tuning",
     "trainTunePyTorch": "PyTorch,Training,Tuning",
-    "trainBenchmark": "PyTorch,Training"
+    "trainBenchmark": "PyTorch,Training",
+    "trainLightning": "Lightning,Training"
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/conf.py:345]: 👌

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "scaling_config = ScalingConfig(\n",
+    "    num_workers=4, use_gpu=False, resources_per_worker={\"CPU\": 1}\n",
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_mnist_example.ipynb:194]: Yeah, just like what we did in the resnet example.

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb (hunk 1 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning_mnist_example.ipynb:54]: Both datamodule and dataloader use a global batch size. PTL will create a distributed sampler for each worker.

**Comment by woshiyyya** [comment]
> Can we also add usage of the `LightningPredictor` at the end? As a separate path to `trainer.test`.

@justinvyu Seems that it's much complex to use `LightningPredictor` than `Trainer.test()`. I am thinking of using the BatchPredictor in the advanced example.


```python
from ray.train.lightning import LightningPredictor
predictor = LightningPredictor.from_checkpoint(checkpoint, MNISTClassifier, use_gpu=False)

def accuracy(logits, labels):
    preds = np.argmax(logits, axis=1)
    correct_preds = np.sum(preds == labels)
    return correct_preds

corrects = total = 0
for batch in test_dataloader:
    inputs, labels = batch
    inputs, labels = inputs.numpy(), labels.numpy()
    logits = predictor.predict(inputs)["predictions"]
    total += labels.size
    corrects += accuracy(logits, labels)
    
print("Accuracy: ", corrects / total)
```

File: doc/BUILD [hunk 6]
```diff
@@ -249,3 +249,13 @@ py_test(
...
+
+py_test(
+    name = "lightning_mnist_example",
+    size = "large",
+    main = "test_myst_doc.py",
+    srcs = ["test_myst_doc.py"],
+    tags = ["exclusive", "team:ml", "gpu", "ray_air"],
+    data = ["//doc/source/train/examples/lightning:lightning_mnist_example"],
+    args = ["--path", "doc/source/train/examples/lightning/lightning_mnist_example.ipynb"]
+)
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/BUILD:261]: Actually these two ipynb files are in different folders. We can't aggregate them in a single build object.

File: doc/source/train/examples/lightning/lightning_mnist_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+    "\n",
+    "The {meth}`LightningConfigBuilder <ray.train.lightning.LightningConfigBuilder>` class stores all the parameters involved in training a PyTorch Lightning module. It takes the same parameter lists as those in PyTorch Lightning.\n",
+    "\n",
+    "- The `.module()` method takes a subclass of `pl.LightningModule` and its initialization parameters. `LightningTrainer` will instantiate a model instance internally in the workers' training loop.\n",
+    "- The `.trainer()` method takes the initialization parameters of `pl.Trainer`. You can specify training configurations, loggers, and callbacks here.\n",
+    "- The `.fit_params()` method stores all the parameters that will be passed into `pl.Trainer.fit()`, including train/val dataloaders, datamodules, and checkpoint paths.\n",
+    "- The `.checkpointing()` method saves the configurations for a `RayModelCheckpoint` callback. This callback reports the latest metrics to the AIR session along with a newly saved checkpoint.\n",
+    "- The `.build()` method generates a dictionary that contains all the configurations in the builder. This dictionary will be passed to `LightningTrainer` later."
```
Comment (richardliaw) [inline doc/source/train/examples/lightning/lightning_mnist_example.ipynb:187]: Can you move this inline into the code cell? Colocating the text with API is easier:

like:

```
lightning_config = (
    LightningConfigBuilder()
    .module(MNISTClassifier, lr=1e-3)  # takes a subclass of `pl.LightningModule` and its initialization parameters.
    .trainer( # takes the initialization parameters of `pl.Trainer. 
        max_epochs=10,
        accelerator="cpu",
        log_every_n_steps=100,
        logger=CSVLogger("logs"),
    )
    .fit_params(datamodule=datamodule)  # stores all the parameters that will be passed into `pl.Trainer.fit()
    .checkpointing(monitor="val_accuracy", mode="max", save_top_k=3)
    .build()
)

**Review by richardliaw — APPROVED** [review]
looks good!

**Pull request merged**